allennlp/models/segmental_language_model.py

```diff
@@ -0,0 +1,300 @@
+    forward direction it should only consider embeddings at previous timesteps, and for
+    its backward direction only embeddings at subsequent timesteps. Similarly, if your
+    language model is unidirectional, the unidirectional contextualizer should only
+    consider embeddings at previous timesteps. If this condition is not met, your
+    language model is cheating.
+
+    Parameters
+    ----------
+    vocab: ``Vocabulary``
+    text_field_embedder: ``TextFieldEmbedder``
+        Used to embed the indexed tokens we get in ``forward``.
+    contextualizer: ``Seq2SeqEncoder``
+        Used to "contextualize" the embeddings. As described above,
+        this encoder must not cheat by peeking ahead.
+    dropout: ``float``, optional (default: None)
+        If specified, dropout is applied to the contextualized embeddings before computation of
+        the softmax. The contextualized embeddings themselves are returned without dropout.
+    num_samples: ``int``, optional (default: None)
+        If provided, the model will use ``SampledSoftmaxLoss``
+        with the specified number of samples. Otherwise, it will use
+        the full ``_SoftmaxLoss`` defined above.
+    sparse_embeddings: ``bool``, optional (default: False)
+        Passed on to ``SampledSoftmaxLoss`` if True.
+    bidirectional: ``bool``, optional (default: False)
+        Train a bidirectional language model, where the contextualizer
+        is used to predict the next and previous token for each input token.
+        This must match the bidirectionality of the contextualizer.
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 contextualizer: Seq2SeqEncoder,
+                 forward_segmental_contextualizer: Seq2SeqEncoder,
+                 backward_segmental_contextualizer: Seq2SeqEncoder,
+                 label_feature_dim: int,
+                 softmax_projection_dim: int,
+                 label_namespace: str = "labels",
+                 dropout: float = None,
+                 num_samples: int = None,
+                 sparse_embeddings: bool = False,
+                 bidirectional: bool = True,
+                 initializer: InitializerApplicator = None) -> None:
+        super().__init__(vocab=vocab,
+                         text_field_embedder=text_field_embedder,
+                         contextualizer=contextualizer,
+                         dropout=dropout,
+                         num_samples=num_samples,
+                         sparse_embeddings=sparse_embeddings,
+                         bidirectional=bidirectional,
+                         initializer=initializer)
+        self._forward_segmental_contextualizer = forward_segmental_contextualizer
+        self._backward_segmental_contextualizer = backward_segmental_contextualizer
+
+        if num_samples is not None:
```

Comment:
I'd add a comment here for why you need a different softmax than the regular LM. I think it's due to the forward dimension size?